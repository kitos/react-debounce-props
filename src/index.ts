import * as React from 'react'
import debounce from 'lodash.debounce'

export interface DebounceProps<T> {
    props: T
    wait: number
    children(attr: T): React.ReactChild | Array<React.ReactChild> | null
}

export class Debounce<T extends Object> extends React.Component<DebounceProps<T>, T> {
    debouncedSetState = debounce(this.setState, this.props.wait)

    componentDidUpdate() {
        this.debouncedSetState(this.props.props)
    }

    render() {
        return this.props.children(this.state || {} as T)
    }
}

export default Debounce
