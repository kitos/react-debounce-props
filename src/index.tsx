import * as React from 'react'
import debounce from 'lodash.debounce'

export interface DebounceProps<T> {
  props: T
  wait: number
  children(attr: T): React.ReactChild | Array<React.ReactChild> | null
}

export class Debounce<T extends Object> extends React.PureComponent<DebounceProps<T>, T> {
  debouncedSetState = debounce(this.setState, this.props.wait)

  componentDidUpdate() {
    this.debouncedSetState(this.props.props)
  }

  render() {
    return this.props.children(this.state || this.props.props)
  }
}

export let withDebouncedProps = (props, wait: number) => Component =>
  class DebounceHOC extends React.PureComponent {
    render() {
      return <Debounce {...{ props, wait }}>{debouncedProps => <Component {...debouncedProps} />}</Debounce>
    }
  }

export default Debounce
