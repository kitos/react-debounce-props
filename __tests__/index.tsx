import * as React from 'react'
import { create } from 'react-test-renderer'
import { Debounce, withDebouncedProps } from '../src'

describe('index', () => {
  describe('render-props', () => {
    it('should render at least', () =>
      expect(
        create(
          <Debounce props={{ test: 42 }} wait={0}>
            {({ test }) => <div>{test}</div>}
          </Debounce>,
        ).toJSON(),
      ).toMatchSnapshot())
  })

  describe('HOC', () => {
    let HOCed = withDebouncedProps({ test: 42 }, 0)(({ test }) => <div>{test}</div>)

    it('should render at least', () => expect(create(<HOCed />).toJSON()).toMatchSnapshot())
  })
})
