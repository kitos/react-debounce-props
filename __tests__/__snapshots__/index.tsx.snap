// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index HOC should render at least 1`] = `
<div>
  42
</div>
`;

exports[`index render-props should render at least 1`] = `
<div>
  42
</div>
`;
